### ISSUE TITLE
High memory usage

### ISSUE THREAD

**Issue by bensheldon** [mperham/dalli#558]
In trying to isolate and reduce the memory usage of my application, I discovered a large difference in memory usage when Dalli is able to connect to memcached  and when it isn't. It seems like when everything is working properly and Dalli is able to connected to the memcached server, it actually uses _more_ memory (up to 50% more memory) than when the memcached server is turned off.  I'm using simple Rails fragment caching in a template.

Is this behavior expected?

Below is the application memory profiling results under the two situations (this is only memory usage of the application, not the memcached server):

```
# Memory usage WITH Memcached server running 
$ derailed exec perf:ram_over_time
PID: 57018
131.09765625
196.51171875
202.265625
203.10546875
222.1953125
237.16796875
266.79296875
226.52734375
228.0390625
223.4921875
285.83203125
264.34765625
237.41796875
287.68359375
283.19921875
285.421875
258.96875
212.25
199.453125
291.7265625
291.84765625
291.87109375
268.59765625

# Memory usage WITHOUT memcached server running
$ derailed exec perf:ram_over_time
PID: 57117
131.15234375
194.34765625
200.484375
201.1953125
201.29296875
201.50390625
202.08984375
202.125
202.13671875
202.2890625
202.30078125
202.39453125
202.40234375
202.421875
202.24609375
177.421875
177.5546875
177.48828125
177.50390625
177.5390625
177.3359375
177.37109375
177.42578125
177.4921875
177.5
177.515625
177.52734375
177.53125
```

Notice that when memcached is _not_ running, the memory usage hovers around 200MB, whereas when memcached _is running_, it jumps around 250MB and goes as high as 290MB.

I'd be happy to provide more information or profiling. Thank you for the help or explanation.

**Comment by petergoldstein** [comment]
I'm not sure I understand the issue. 

When used as a fragment cache the purpose of Memcached / Dalli isn't to minimize memory usage, it's to minimize render time for pages.  Depending on the nature of the application and how Dalli is being used, it's perfectly reasonable that you might see higher memory usage when Memcached is not available.

**Comment by bensheldon** [comment]
@petergoldstein that's the explanation I was looking for because I knew that the purpose of memcached is to cut down on render time, not necessarily memory. But I was rather surprised to see Dalli+memcached, when working correctly, increasing the amount of memory my application used by 50% as well as jumping around so much in the amount of memory used.

...but if that's not unexpected from your perspective, I'll accept that.

**Comment by bensheldon** [comment]
![screen shot 2015-09-11 at 7 02 20 pm](https://cloud.githubusercontent.com/assets/47554/9829410/ac7c47ac-58b7-11e5-8ef4-e5497f89aa7c.png)

Just some further data. Just to experiment, I disabled our production memcached server to see what the effect was. As you can see, the total memory dropped by about 40%, whereas object allocations and GC increased (because of all of the template rendering that is now having to happen). 

I just find it surprising that essentially the sum of all those individual template strings, when retrieved from memcached, uses so much more memory than rendering them individually.

**Comment by jjb** [comment]
@petergoldstein: @bensheldon is talking about memory used by his app, not by the machine that memcached is on. Why would using dalli significantly increase the amount of memory used by the app?

**Comment by petergoldstein** [comment]
@jjb Yes, I understand that, and that was the situation I was addressing.  As per my earlier response, the purpose of using Dalli / Memcached is to minimize render time for pages - the latency of the individual page request.  Minimizing memory usage is not the goal.  So as part of that process it may create a number of object that are long lived, and would substantially increase memory usage.

I won't deny that an additional 100MB for Dalli/Memcached seems high, but without going into a detailed analysis of his particular application, I can't speak in any more detail than that.

So unless there's evidence of a memory leak (not indicated in the data above), it's simply not an issue for Dalli.

**Comment by jjb** [comment]
Gotcha.  

@bensheldon your screenshot only shows the point of restart without dalli. After you restart your app when using dalli, how much memory is it initially taking?

**Comment by bensheldon** [comment]
@jjb my [first comment](https://github.com/mperham/dalli/issues/558#issue-106121958) in this thread has raw profiling numbers. 130MB at app startup; 200MB on 1st request, and then the divergence.

Also to clarify, the comparisons (both the numbers from 1st comment and New Relic charts) is not "with/without dalli", but "dalli that can/cannot connect memcached server". In the comparisons, I'm not removing Dalli from my app, but rather just disabling the memcached server, so that Dalli is still _trying_ to connect, but unable to.

Back when I filed this issue, I spent more time profiling, and it seemed like the memory usage was spiking when Dalli is reading out the memcached socket connection. (I was experimenting with caching a single template with 1MB of lorem-ipsum text and still saw high memory usage)  I also tried using the `kgio` gem and still saw the high memory usage with that as well. I ultimately gave up and am using Rails' :memory_cache, which uses less memory overall (even with multiple puma processes and threads) than when using Dalli (which is counterintuitive, I know). I'd be happy to help profile further if you think it would be helpful; it overall seemed like I was having a non-normative experience with Dalli.
